ouroboros-consensus/test-consensus/Test/Dynamic/Util/Expectations.hs

```diff
@@ -80,43 +86,83 @@ data Acc = Acc
+    prj $ foldl' step initial (Map.toAscList sched)
   where
     prj Acc{maxForkLength} = NumBlocks maxForkLength
 
     -- assume the network begins in consensus (eg all nodes start at Genesis)
     initial = Acc
-      { maxForkLength = 0
+      { maxChainLength = 0
+      , maxForkLength = 0
       }
 
     -- this logic focuses on the new chains made in this slot that are longer
     -- than the longest chains from the previous slot
-    step Acc{maxForkLength} (_slot, leaders) =
+    step Acc{maxChainLength, maxForkLength} (slot, leaders) =
         Acc
-          { maxForkLength = update maxForkLength
+          { maxChainLength = grow    maxChainLength
+          , maxForkLength  = update  maxForkLength
           }
       where
+        grow = if 0 == pullingAhead then id else (+ 1)
+
         update
             -- too late to reach consensus, so further diverge
-          | maxForkLength > maxRollbacks k = (+ 1)
+          | maxForkLength > maxRollbacks k = grow
 
             -- assume (common) worst-case: each leader creates a unique longer
             -- chain
-          | numLeaders >  1                = (+ 1)
+          | pullingAhead >  1              = (+ 1)
 
             -- the sole leader creates the sole longer chain, bringing the
             -- network into consensus
-          | numLeaders == 1                = const 0
+          | pullingAhead == 1              = const 0
 
-            -- no leaders, so the longest chains are the same as in the
-            -- previous slot
+            -- there will be multiple longest chains that disagree on at least
+            -- the latest block
+            --
+            -- Note that pullingAhead == 0 by the preceding guards
+          | pullingEven  >  0              = max 1
+
+            -- no nodes are extending their chain, so the longest chains are
+            -- the same as in the previous slot
           | otherwise                      = id
 
-        -- ASSUMPTION: each leader is making a unique new chain that is longer
-        -- than any from the previous slot
-        numLeaders = length leaders
+        -- how many leaders are forging a block onto a @maxForkLength@-chain
+        pullingAhead = nlOld + nlNew (maxChainLength == 0)
+        -- how many leaders are forging a block onto a @maxForkLength - 1@-chain
+        pullingEven  = nlNew (maxChainLength == 1)
+
+        -- how many leaders joined before this slot
+        nlOld = length $ filter ((< slot) . joinSlot) leaders
+        nlNew b
+          | not b     = 0
+            -- how many leaders are joining during this slot; these might not
+            -- be relevant
+            --
+            -- A node leading the same slot it joins always forges a 1-block
+            -- chain; there's actually a race-condition between blocking
+            -- forging and BlockFetch/ChainSync, but forging always wins in the
+            -- current test framework implementation.
```

Comment:
I think those delays will also have to be passed in to this function, like `LeaderSchedule`, `NodeJoinPlan`, etc. So this `determineExpectations` function is likely to get mired by a lot of details. It'll still be simpler than the implementation of course, but it will become much less simple of a model when we start anticipating the consequences of topologies, temporary network partitions, etc.